File: ts/test/review.test.ts
/* Copyright (c) 2021-2024 Richard Rodger and other contributors, MIT License */

// Behaviour the documentation review of 2026-09-02 found wrong or unpinned.

import { describe, test } from 'node:test'
import assert from 'node:assert'

import { Shape as ShapeX } from '../dist/shape'

let ShapeModule = require('../dist/shape')
if (ShapeModule.Shape) {
  ShapeModule = ShapeModule.Shape
}
const Shape: ShapeX = ShapeModule
const { buildize, Min, Max, Some, One, All, Open, Coerce, fromJsonSchema } = ShapeModule

describe('review', () => {
  test('a chained Rest keeps the tuple', () => {
    const s = Shape(buildize([String, Number]).Rest(Number))
    assert.deepEqual(s(['a', 1, 2, 3]), ['a', 1, 2, 3])
    assert.throws(() => s(['a', 1, 'b']), /index "2" with string "b"/)
    assert.throws(() => s([1, 1]), /index "0" with number "1"/)
    // The top-level form is the same shape.
    assert.deepEqual(Shape(ShapeModule.Rest(Number, [String, Number]))(['a', 1, 2]), ['a', 1, 2])
    // Chained on a non-array node it is a plain element shape, as before.
    assert.deepEqual(Shape(ShapeModule.Optional().Rest(Number))([1, 2]), [1, 2])
  })

  test('pathArr under an array element', () => {
    const errs = Shape({ users: [{ email: String }] }).error({ users: [{ email: 1 }] })
    assert.deepEqual(errs.map(e => [e.path, e.pathArr]), [['users.0.email', ['users', 0, 'email']]])
    assert.deepEqual(Shape([[{ x: Number }]]).error([[{ x: 's' }]]).map(e => e.pathArr), [[0, 0, 'x']])
    assert.deepEqual(Shape({ a: { b: [String] } }).error({ a: { b: [1] } }).map(e => e.pathArr), [['a', 'b', 0]])
    assert.deepEqual(Shape([String]).error([1]).map(e => e.pathArr), [[0]])
    assert.deepEqual(Shape({ a: String }).error({ a: 1 }).map(e => e.pathArr), [['a']])
  })

  test('a bound on a missing required value defers to the missing error', () => {
    assert.throws(() => Shape({ a: Min(2, String) })({}), /property "a" because the property is missing\.$/)
    assert.throws(() => Shape({ a: Max(2, Number) })({}), /property "a" because the property is missing\.$/)
    // With a default the bound never sees the absence.
    assert.deepEqual(Shape({ a: Min(2, 'abc') })({}), { a: 'abc' })
  })

  test('One, Some and All never change the value they are given', () => {
    // Every Some branch sees the original; the last matching result stands.
    const input = {}
    assert.deepEqual(Shape(Some(Open({ a: 1 }), Open({ a: 2 })))(input), { a: 2 })
    assert.deepEqual(input, {})
    assert.deepEqual(Shape(Some(Open({ a: 1 }), Open({ b: 2 })))({}), { b: 2 })
    assert.equal(Shape(Some(Coerce(Number), Max(2)))('12'), '12')
    assert.equal(Shape(Some(Max(2), Coerce(Number)))('12'), 12)
    assert.deepEqual(Shape(Some(Coerce(Number), Open({ a: 1 })))({}), { a: 1 })
    // A failing branch leaks nothing into the value.
    assert.deepEqual(Shape(Some({ x: Number, y: 1 }, Open({ z: 3 })))({}), { z: 3 })
    // All threads a copy; a failing composition leaves the input untouched.
    const inner = {}
    const parent = { x: inner }
    assert.deepEqual(Shape({ x: All(Open({ a: 1 }), Open({ b: 2 })) })(parent), { x: { a: 1, b: 2 } })
    assert.deepEqual(inner, {})
    assert.equal(Shape(All(Coerce(Number), Min(2)))('5'), 5)
    const failing = {}
    assert.throws(() => Shape(All(Open({ a: 1 }), { z: String }))(failing), /does not satisfy all of/)
    assert.deepEqual(failing, {})
    // One produces from a copy too.
    const one = {}
    assert.deepEqual(Shape(One(Open({ a: 1 }), String))(one), { a: 1 })
    assert.deepEqual(one, {})
  })

  test('JSON Schema import keeps a plain bound beside an exclusive one', () => {
    const lo = Shape(fromJsonSchema({ type: 'number', minimum: 1, exclusiveMinimum: 0 }))
    assert.equal(lo(1), 1)
    assert.throws(() => lo(0.5), /minimum of 1/)
    const hi = Shape(fromJsonSchema({ type: 'number', maximum: 1, exclusiveMaximum: 2 }))
    assert.equal(hi(1), 1)
    assert.throws(() => hi(1.5), /maximum of 1/)
  })

  test('error() returns issue objects', () => {
    const errs = Shape({ a: Number }).error({ a: 'x' })
    assert.equal(errs.length, 1)
    assert.equal(errs[0].path, 'a')
    assert.equal(errs[0].why, 'type')
    assert.equal(errs[0].value, 'x')
  })
})

File: go/review_test.go
package shape

import (
	"math"
	"regexp"
	"testing"
)

// Behaviour the documentation review of 2026-09-02 found wrong or unpinned.

// The string form rejects a builder called wrongly, as expr throws in
// TypeScript.
func TestExprFaultIsAnError(t *testing.T) {
	for src, want := range map[string]string{
		`String.Min("x")`:  "Shape: Min needs a number",
		`Len(-1)`:          "Shape: Len needs a whole number of zero or more",
		`Open(Define(""))`: "Shape: Define needs a name",
	} {
		if _, err := Expr(src); err == nil || err.Error() != want {
			t.Fatalf("%s: got %v, want %s", src, err, want)
		}
	}
}

// Exact compares numbers by value across kinds, so an int literal in the
// spec matches the float64 a JSON decoder produces.
func TestExactNumberKinds(t *testing.T) {
	s, err := Shape(Exact(1, "a"))
	if err != nil {
		t.Fatal(err)
	}
	for _, in := range []any{1, 1.0, int64(1), int8(1), uint(1), float32(1), "a"} {
		mustValid(t, s, in)
	}
	for _, in := range []any{2, 1.5, "1", true, nil} {
		if _, err := s.Validate(in); err == nil {
			t.Fatalf("%v: expected an error", in)
		}
	}
	// Large integers keep their precision: no detour through float64.
	big, _ := Shape(Exact(int64(9007199254740993)))
	mustValid(t, big, int64(9007199254740993))
	mustValid(t, big, uint64(9007199254740993))
	for _, in := range []any{int64(9007199254740992), float64(9007199254740992), uint64(9007199254740994)} {
		if _, err := big.Validate(in); err == nil {
			t.Fatalf("%v: expected an error", in)
		}
	}
	// NaN equals nothing, as in JavaScript.
	nan, _ := Shape(Exact(math.NaN()))
	if _, err := nan.Validate(math.NaN()); err == nil {
		t.Fatal("NaN: expected an error")
	}
	if exactRat("x") != nil {
		t.Fatal("exactRat of a string")
	}
}

// A deliberate Fault on a Never node is a valid expression; only a builder
// given a wrong argument is refused.
func TestExprKeepsDeliberateFaults(t *testing.T) {
	for _, src := range []string{`Never.Fault("f")`, `Fault("f", Never)`} {
		n, err := Expr(src)
		if err != nil {
			t.Fatalf("%s: %v", src, err)
		}
		s, _ := Shape(n)
		if _, err := s.Validate(1); err == nil || err.Error() != "f" {
			t.Fatalf("%s: got %v, want f", src, err)
		}
	}
}

// A Check is not called for an absent value; the required check speaks.
func TestCheckSkipsAbsent(t *testing.T) {
	called := false
	s, _ := Shape(map[string]any{"a": Check(func(val any, update *Update, state *State) bool {
		called = true
		return true
	})})
	want := `Validation failed for property "a" because the property is missing.`
	if _, err := s.Validate(map[string]any{}); called || err == nil || err.Error() != want {
		t.Fatalf("called=%v err=%v", called, err)
	}
	re, _ := Shape(map[string]any{"a": Check(regexp.MustCompile("^x"))})
	if _, err := re.Validate(map[string]any{}); err == nil || err.Error() != want {
		t.Fatalf("regexp: %v", err)
	}
}

// Every Some branch sees the value as given, never one another branch
// changed; the last matching branch's result stands, and the subject is
// not changed.
func TestSomeBranchesSeeTheOriginal(t *testing.T) {
	s, _ := Shape(Some(Open(map[string]any{"a": 1}), Open(map[string]any{"a": 2})))
	in := map[string]any{}
	out := mustValid(t, s, in).(map[string]any)
	if out["a"] != 2 || len(in) != 0 {
		t.Fatalf("got %v, in %v, want a: 2 and an untouched input", out, in)
	}
	both, _ := Shape(Some(Open(map[string]any{"a": 1}), Open(map[string]any{"b": 2})))
	if out := mustValid(t, both, map[string]any{}).(map[string]any); out["b"] != 2 || out["a"] != nil {
		t.Fatalf("got %v, want b: 2 only", out)
	}
	sc, _ := Shape(Some(Coerce(Number), Max(2)))
	if out := mustValid(t, sc, "12"); out != "12" {
		t.Fatalf("got %v (%T), want \"12\"", out, out)
	}
	cs, _ := Shape(Some(Max(2), Coerce(Number)))
	if out := mustValid(t, cs, "12"); out != 12.0 {
		t.Fatalf("got %v (%T), want 12", out, out)
	}
	// A branch that replaces the value leaves the next branch the original.
	rp, _ := Shape(Some(Catch(1, Number), Open(map[string]any{"a": 1})))
	if out := mustValid(t, rp, map[string]any{}).(map[string]any); out["a"] != 1 {
		t.Fatalf("got %v, want a: 1", out)
	}
	if out := mustValid(t, rp, 3); out != 3.0 && out != 3 {
		t.Fatalf("got %v (%T), want 3", out, out)
	}
	// All threads a copy; a failing composition leaves the input untouched.
	all, _ := Shape(All(Open(map[string]any{"a": 1}), Open(map[string]any{"b": 2})))
	if out := mustValid(t, all, map[string]any{}).(map[string]any); out["a"] != 1 || out["b"] != 2 {
		t.Fatalf("got %v, want a: 1, b: 2", out)
	}
	bad, _ := Shape(All(Open(map[string]any{"a": 1}), map[string]any{"z": String}))
	in2 := map[string]any{}
	if _, err := bad.Validate(in2); err == nil || len(in2) != 0 {
		t.Fatalf("err %v, in %v", err, in2)
	}
}
